### PR TITLE
Fix for an issue with redeferral and stack func init in jitted code

### PR DESCRIPTION
**Pull request by rajatd**
For initializing a nested function object on the stack, the jitted code was using hardcoded field values and offsets from the function's functionProxy. This is not legal with redeferral as the proxy could have been a full function body and could be redeferred from the time its enclosing function was jitted to when it was executed. Redeferral will throw away the existing function body and reallocate a parseable function info which will then be set as function proxy on the function info. This reallocation means that the hardcoded values and offsets in the jitted code are no longer valid.

Fix is to not use compile time values but to jit indirections. We already do that for oop jit, just made in-proc jit also do the same.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/2322)

### CODE REVIEW

File: lib/Backend/Lower.cpp [hunk 1]
```diff
@@ -6579,73 +6579,40 @@ Lowerer::GenerateScriptFunctionInit(IR::RegOpnd * regOpnd, IR::Opnd * vtableAddr
...
-        {
-            functionInfoOpnd = IR::IndirOpnd::New(functionProxyOpnd->AsRegOpnd(), Js::FunctionProxy::GetOffsetOfFunctionInfo(), TyMachReg, func);
-        }
-        else
-        {
-            functionInfoOpnd = IR::MemRefOpnd::New((BYTE*)functionProxyOpnd->AsAddrOpnd()->m_address + Js::FunctionProxy::GetOffsetOfFunctionInfo(), TyMachReg, func);
-        }
-    }
+    
+    functionInfoOpnd = IR::IndirOpnd::New(functionProxyOpnd->AsRegOpnd(), Js::FunctionProxy::GetOffsetOfFunctionInfo(), TyMachReg, func);
```
Comment (rajatd) [inline lib/Backend/Lower.cpp]: @pleath - i was in two minds whether to directly use the nestedInfo for initializing functionInfo. I'm sure I have asked you this before, but do we expect functionInfo to be different from functionInfo->functionBodyImpl->functionInfo in some cases?

File: lib/Backend/Lower.cpp [hunk 2]
```diff
@@ -6579,73 +6579,40 @@ Lowerer::GenerateScriptFunctionInit(IR::RegOpnd * regOpnd, IR::Opnd * vtableAddr
...
     GenerateMemInitNull(regOpnd, Js::ScriptFunction::GetOffsetOfObjectArray(), insertBeforeInstr, isZeroed);
     GenerateMemInit(regOpnd, Js::ScriptFunction::GetOffsetOfConstructorCache(),
-        LoadLibraryValueOpnd(insertBeforeInstr, LibraryValue::ValueConstructorCacheDefaultInstance),
+    LoadLibraryValueOpnd(insertBeforeInstr, LibraryValue::ValueConstructorCacheDefaultInstance),
         insertBeforeInstr, isZeroed);
-    if (!functionInfoOpnd)
-    {
-        if (functionProxyOpnd->IsRegOpnd())
-        {
-            functionInfoOpnd = IR::IndirOpnd::New(functionProxyOpnd->AsRegOpnd(), Js::FunctionProxy::GetOffsetOfFunctionInfo(), TyMachReg, func);
```
Comment (ThomsonTan) [inline lib/Backend/Lower.cpp]: >functionInfoOpnd [](start = 12, length = 16)

I am also wondering why we need re-check functionInfo from functionProxy, should it be the same?

**Comment by rajatd** [comment]
@pleath please review

File: lib/Backend/Lower.cpp (hunk 1 above)
Comment (pleath) [inline lib/Backend/Lower.cpp]: No. Never. Once we allocate a FunctionInfo for a given user function, it never changes. So, yes, once we've loaded it from the nested array we can just reuse it throughout the code sequence.

**Comment by dilijev** [comment]
You can now resolve the CI issues as described here: https://github.com/Microsoft/ChakraCore/pull/2332#issuecomment-271029950

**Comment by dilijev** [comment]
@dotnet-bot test this please

**Comment by pleath** [comment]
Change checked in under another PR.

**Pull request closed without merging**